ex19/doc/Wsel.3
'\" t
.TH "" 3 "" "Version Unknown To Mankind" "Free Widget Foundation"
.SH NAME
Wsel
.SH DESCRIPTION

.SS "Public variables"

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Wsel
Name	Class	Type	Default
XtNvalue	XtCValue	int 	0 
XtNlocked	XtCLocked	Bool 	0 
XtNnames	XtCNames	StringMArray 	0 

.TE
.ps +2


.TP
.I "XtNvalue"



.TP
.I "XtNlocked"



.TP
.I "XtNnames"



.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Wlab
Name	Class	Type	Default
XtNlabel	XtCLabel	String 	NULL 
XtNcornerRoundPercent	XtCCornerRoundPercent	Int 	0 
XtNinnerBorder	XtCInnerBorder	Int 	0 
XtNupdate	XtCUpdate	Int 	0 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Wheel
Name	Class	Type	Default
XtNxftFont	XtCXFtFont	XftFont	"Sans-22"
XtNcallback	XtCCallback	Callback	NULL 
XtNbg_norm	XtCBg_norm	Pixel	"lightblue"
XtNbg_sel	XtCBg_sel	Pixel	"yellow"
XtNbg_hi	XtCBg_hi	Pixel	"red"
XtNfg_norm	XtCFg_norm	Pixel	"black"
XtNfg_sel	XtCFg_sel	Pixel	"green"
XtNfg_hi	XtCFg_hi	Pixel	"white"
XtNuser_data	XtCUser_data	Int 	0 
XtNfocus_group	XtCFocus_group	String 	""
XtNstate	XtCState	Int 	0 
XtNregister_focus_group	XtCRegister_focus_group	Boolean 	True 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Core
Name	Class	Type	Default
XtNx	XtCX	Position 	0 
XtNy	XtCY	Position 	0 
XtNwidth	XtCWidth	Dimension 	0 
XtNheight	XtCHeight	Dimension 	0 
borderWidth	XtCBorderWidth	Dimension 	0 
XtNcolormap	XtCColormap	Colormap 	NULL 
XtNdepth	XtCDepth	Int 	0 
destroyCallback	XtCDestroyCallback	XTCallbackList 	NULL 
XtNsensitive	XtCSensitive	Boolean 	True 
XtNtm	XtCTm	XTTMRec 	NULL 
ancestorSensitive	XtCAncestorSensitive	Boolean 	False 
accelerators	XtCAccelerators	XTTranslations 	NULL 
borderColor	XtCBorderColor	Pixel 	0 
borderPixmap	XtCBorderPixmap	Pixmap 	NULL 
background	XtCBackground	Pixel 	0 
backgroundPixmap	XtCBackgroundPixmap	Pixmap 	NULL 
mappedWhenManaged	XtCMappedWhenManaged	Boolean 	True 
XtNscreen	XtCScreen	Screen *	NULL 

.TE
.ps +2

.SS "Translations"


.nf
<EnterWindow>: highlight() 
.fi



.nf
<LeaveWindow>: reset() 
.fi



.nf
<BtnDown>: notify() 
.fi



.nf
<Btn1Motion>: motion_start() 
.fi


.SS "Actions"


.TP
.I "highlight




.TP
.I "reset




.TP
.I "notify




.TP
.I "motion_start





ex19/src/Wsel.c
/* Generated by wbuild
 * (generator version 3.3)
 */
#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#line 108 "Wsel.widget"
#include "mls.h"
#line 109 "Wsel.widget"
#include <X11/Xft/Xft.h>
#line 110 "Wsel.widget"
#include "converters.h"
#include <xtcw/WselP.h>
#line 57 "Wsel.widget"
static void highlight(
#if NeedFunctionPrototypes
Widget,XEvent*,String*,Cardinal*
#endif
);
#line 63 "Wsel.widget"
static void reset(
#if NeedFunctionPrototypes
Widget,XEvent*,String*,Cardinal*
#endif
);
#line 68 "Wsel.widget"
static void notify(
#if NeedFunctionPrototypes
Widget,XEvent*,String*,Cardinal*
#endif
);
#line 76 "Wsel.widget"
static void motion_start(
#if NeedFunctionPrototypes
Widget,XEvent*,String*,Cardinal*
#endif
);

static XtActionsRec actionsList[] = {
{"highlight", highlight},
{"reset", reset},
{"notify", notify},
{"motion_start", motion_start},
};

static char defaultTranslations[] = "\
<EnterWindow>: highlight() \n\
<LeaveWindow>: reset() \n\
<BtnDown>: notify() \n\
<Btn1Motion>: motion_start() \n\
";
static void _resolve_inheritance(
#if NeedFunctionPrototypes
WidgetClass
#endif
);
#line 19 "Wsel.widget"
static void initialize(
#if NeedFunctionPrototypes
Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 26 "Wsel.widget"
static Boolean  set_values(
#if NeedFunctionPrototypes
Widget ,Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 34 "Wsel.widget"
static void destroy(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 41 "Wsel.widget"
static String  get_name(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 41 "Wsel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 41 "Wsel.widget"
static String  get_name(Widget self)
#else
#line 41 "Wsel.widget"
static String  get_name(self)Widget self;
#endif
#line 42 "Wsel.widget"
{
  if( ((WselWidget)self)->wsel.max_value == 0 ) return "";
  if( ((WselWidget)self)->wsel.value < 0 ) ((WselWidget)self)->wsel.value = ((WselWidget)self)->wsel.max_value -1;
  else if( ((WselWidget)self)->wsel.value >= ((WselWidget)self)->wsel.max_value ) ((WselWidget)self)->wsel.value = 0;
  return STR(((WselWidget)self)->wsel.names,((WselWidget)self)->wsel.value);
}

static XtResource resources[] = {
#line 7 "Wsel.widget"
{XtNvalue,XtCValue,XtRInt,sizeof(((WselRec*)NULL)->wsel.value),XtOffsetOf(WselRec,wsel.value),XtRImmediate,(XtPointer)0 },
#line 8 "Wsel.widget"
{XtNlocked,XtCLocked,XtRBool,sizeof(((WselRec*)NULL)->wsel.locked),XtOffsetOf(WselRec,wsel.locked),XtRImmediate,(XtPointer)0 },
#line 9 "Wsel.widget"
{XtNnames,XtCNames,XtRStringMArray,sizeof(((WselRec*)NULL)->wsel.names),XtOffsetOf(WselRec,wsel.names),XtRImmediate,(XtPointer)0 },
};

WselClassRec wselClassRec = {
{ /* core_class part */
/* superclass   	*/  (WidgetClass) &wlabClassRec,
/* class_name   	*/  "Wsel",
/* widget_size  	*/  sizeof(WselRec),
/* class_initialize 	*/  NULL,
/* class_part_initialize*/  _resolve_inheritance,
/* class_inited 	*/  FALSE,
/* initialize   	*/  initialize,
/* initialize_hook 	*/  NULL,
/* realize      	*/  XtInheritRealize,
/* actions      	*/  actionsList,
/* num_actions  	*/  4,
/* resources    	*/  resources,
/* num_resources 	*/  3,
/* xrm_class    	*/  NULLQUARK,
/* compres_motion 	*/  False ,
/* compress_exposure 	*/  XtExposeCompressSeries ,
/* compress_enterleave 	*/  False ,
/* visible_interest 	*/  True ,
/* destroy      	*/  destroy,
/* resize       	*/  XtInheritResize,
/* expose       	*/  XtInheritExpose,
/* set_values   	*/  set_values,
/* set_values_hook 	*/  NULL,
/* set_values_almost 	*/  XtInheritSetValuesAlmost,
/* get_values+hook 	*/  NULL,
/* accept_focus 	*/  XtInheritAcceptFocus,
/* version      	*/  XtVersion,
/* callback_private 	*/  NULL,
/* tm_table      	*/  defaultTranslations,
/* query_geometry 	*/  XtInheritQueryGeometry,
/* display_acceleator 	*/  XtInheritDisplayAccelerator,
/* extension    	*/  NULL 
},
{ /* Wheel_class part */
XtInherit_exec_command,
XtInherit_sig_recv,
},
{ /* Wlab_class part */
XtInherit_cache_hit,
XtInherit_minimal_size,
XtInherit_prefered_size,
},
{ /* Wsel_class part */
 /* dummy */  0
},
};
WidgetClass wselWidgetClass = (WidgetClass) &wselClassRec;
/*ARGSUSED*/
#line 57 "Wsel.widget"
static void highlight(self,event,params,num_params)Widget self;XEvent*event;String*params;Cardinal*num_params;
{
  if( ((WselWidget)self)->wsel.locked ) return;
  XtVaSetValues(self,"state", STATE_SELECTED, NULL );
}

/*ARGSUSED*/
#line 63 "Wsel.widget"
static void reset(self,event,params,num_params)Widget self;XEvent*event;String*params;Cardinal*num_params;
{
    XtVaSetValues(self,"state", STATE_NORMAL, NULL );
}

/*ARGSUSED*/
#line 68 "Wsel.widget"
static void notify(self,event,params,num_params)Widget self;XEvent*event;String*params;Cardinal*num_params;
{
  if( ((WselWidget)self)->wsel.locked ) return;
  ((WselWidget)self)->wsel.value = ((WselWidget)self)->wsel.value +1;
  XtVaSetValues(self,"label", get_name(self), NULL );
  XtCallCallbackList( self, ((WselWidget)self)->wheel.callback, (XtPointer) ((WselWidget)self)->wsel.value );
}

/*ARGSUSED*/
#line 76 "Wsel.widget"
static void motion_start(self,event,params,num_params)Widget self;XEvent*event;String*params;Cardinal*num_params;
{
  if( ((WselWidget)self)->wsel.locked ) return;

	int y = event->xbutton.y;
        int x = event->xbutton.x;
	ulong t = (ulong ) event->xbutton.time;

	/* drag startete vor über einer sekunde, seitdem keine änderung */
	/* darum: vergessen und neu starten */
	if( ((WselWidget)self)->wsel.drag_update + 1000 < t ) goto restart_drag;
	((WselWidget)self)->wsel.drag_update = t;

        int move = (((WselWidget)self)->wsel.drag_last_y - y) / 30;
        if(!move) move = (((WselWidget)self)->wsel.drag_last_x - x) / 30;
        if( move ) {
            ((WselWidget)self)->wsel.drag_last_x = x;
	    ((WselWidget)self)->wsel.drag_last_y = y;
            printf("Move %d\n", move );
            ((WselWidget)self)->wsel.value = ((WselWidget)self)->wsel.value +move;
            XtVaSetValues(self,"label", get_name(self), NULL );
        }
	return; /* drag in progress */

  restart_drag:
	((WselWidget)self)->wsel.drag_last_y = event->xbutton.y;
        ((WselWidget)self)->wsel.drag_last_x = event->xbutton.x;
	((WselWidget)self)->wsel.drag_update = t;
}

static void _resolve_inheritance(class)
WidgetClass class;
{
  WselWidgetClass c __attribute__((unused)) = (WselWidgetClass) class;
  WselWidgetClass super __attribute__((unused));
  if (class == wselWidgetClass) return;
  super = (WselWidgetClass)class->core_class.superclass;
}
#line 19 "Wsel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 19 "Wsel.widget"
static void initialize(Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 19 "Wsel.widget"
static void initialize(request,self,args,num_args)Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 20 "Wsel.widget"
{
        if( ((WselWidget)self)->wsel.names == 0 ) ((WselWidget)self)->wsel.max_value=0;
        else ((WselWidget)self)->wsel.max_value = m_len(((WselWidget)self)->wsel.names);
        ((WselWidget)self)->wlab.label = get_name(self);
}
#line 26 "Wsel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 26 "Wsel.widget"
static Boolean  set_values(Widget  old,Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 26 "Wsel.widget"
static Boolean  set_values(old,request,self,args,num_args)Widget  old;Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 27 "Wsel.widget"
{
        if( ((WselWidget)old)->wsel.value != ((WselWidget)self)->wsel.value ) {
            XtVaSetValues(self,"label", get_name(self), NULL );
        }
        return 0;
}
#line 34 "Wsel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 34 "Wsel.widget"
static void destroy(Widget self)
#else
#line 34 "Wsel.widget"
static void destroy(self)Widget self;
#endif
#line 35 "Wsel.widget"
{
        m_free_strings(((WselWidget)self)->wsel.names,0);
}
